### PR TITLE
Export gcode

### PR DESCRIPTION
**Pull request by mrzl**
Ok, this is a beginning, to get the stone rolling. I've been implementing some things according to your advice https://github.com/t-oster/LibLaserCut/issues/32

The commits for LibLaserCut are in the separate repository, for suggestions on how to handle this, let me know.

### CODE REVIEW

File: src/com/t_oster/visicut/VisicutModel.java [hunk 1]
```diff
@@ -757,6 +757,24 @@ public void sendJob(String name, ProgressListener pl, Map<LaserProfile, List<Las
     }
   }
 
+  public void saveJob( String name, ProgressListener pl, Map<LaserProfile, List<LaserProperty>> props, List<String> warnings ) throws IllegalJobException, SocketTimeoutException, Exception{
+	  LaserCutter lasercutter = this.getSelectedLaserDevice().getLaserCutter();
```
Comment (t-oster) [inline src/com/t_oster/visicut/VisicutModel.java]: As the counterpart in LibLaserCut, this method should get an output stream where to put the data.

File: src/com/t_oster/visicut/gui/MainView.java [hunk 2]
```diff
@@ -1789,7 +1798,96 @@ private void editMappings() throws FileNotFoundException, IOException
...
-
+	private void generateGcodeMenuItemActionPerformed( java.awt.event.ActionEvent evg ) {
+		String jobname = getJobName();
+		List<String> warnings = new LinkedList<String>();
+		final Map<LaserProfile, List<LaserProperty>> cuttingSettings = this.getPropertyMapForCurrentJob();
+		if (cuttingSettings == null)
+		{
+			return;
+		}
+
```
Comment (t-oster) [inline src/com/t_oster/visicut/gui/MainView.java]: Here we should ask the user, where to save the file. One thing, we did not consider yet is the file extension (not every driver outputs gcode). To make it perfect, the driver should give the file extension (e.g. via a method getOutputFileExtension())... but right now I could live with ".dat" or something like this.

**Comment by mrzl** [comment]
Cheers, I adopted your thoughts, and went the easy way of throwing an exception, when a lasercutter is used that doesn't implement the saveJob() method for now.

**Comment by mrzl** [comment]
By the way, a huge advantage of being able to export gcode, at least for me, using a smoothieboard, is that you can decrease the time a cutting job takes massively by uploading the generated gcode via the webinterface of the smoothieboard. It then doesn't wait for the 'ok' reply after every command. But I recon that behavior can be achieved by setting setWaitForOKafterEachLine to false, right?

**Comment by t-oster** [comment]
The driver is able to upload the file via Web interface, too. Just unset Hostname and port and let the http upload url be correct. 
I think we need a wiki page for the driver

-------- Ursprüngliche Nachricht --------
Von: Marcel Schwittlick <notifications@github.com> 
Datum:29.11.2015  03:59  (GMT+01:00) 
An: t-oster/VisiCut <VisiCut@noreply.github.com> 
Cc: Thomas Oster <mail@thomas-oster.de> 
Betreff: Re: [VisiCut] Export gcode (#316) 

By the way, a huge advantage of being able to export gcode, at least for me, using a smoothieboard, is that you can decrease the time a cutting job takes massively by uploading the generated gcode via the webinterface of the smoothieboard. It then doesn't wait for the 'ok' reply after every command. But I recon that behavior can be achieved by setting setWaitForOKafterEachLine to false, right?

**Pull request merged**